rpk: add `rpk topic trim-prefix` (support for DeleteRecords)

r-vasquez opened the pull request:
This PR introduces the new command `rpk topic trim-prefix` that supports the DeleteRecords admin request

Fixes #11781

### Usage and Help Text
**Full Help Text**
```
$ rpk topic trim-prefix --help
Trim records from topics

This command allows you to trim records from topics, to trim the topics Redpanda
sets the LogStartOffset for partitions to the requested offset. All segments 
whose max partition is before the requested offset are deleted, and any records 
within the segment before the requested offset can no longer be read.

The --offset/-o flag allows you to indicate which index you want to set the 
partition's low watermark (start offset) to. It can be a single integer value 
denoting the offset or a timestamp if you prefix the offset with an '@'. You may
select which partition you want to trim the offset from with the --partition/-p 
flag.

The --from-file option allows to trim the offsets specified in a text file with
the following format:
    [TOPIC] [PARTITION] [OFFSET]
    [TOPIC] [PARTITION] [OFFSET]
    ...
or the equivalent keyed JSON/YAML file.

EXAMPLES

Trim records in 'foo' topic to offset 120 in partition 1
    rpk topic trim-prefix foo --offset 120 --partition 1

Trim records in all partitions of topic foo previous to an specific timestamp
    rpk topic trim-prefix foo -o "@1622505600"

Trim records from a JSON file
    rpk topic trim-prefix --from-file /tmp/to_trim.json

Usage:
  rpk topic trim-prefix [TOPIC] [flags]

Aliases:
  trim-prefix, trim

Flags:
  -f, --from-file string   File of topic/partition/offset for which to trim offsets for
  -h, --help               Help for trim-prefix
      --no-confirm         Disable confirmation prompt
  -o, --offset string      Offset to set the partition's low watermark (start offset) to
  -p, --partitions ints    Comma-separated list of partitions to trim records from (default to all)
      --to-hwm             Set the partition's low watermark to the partition's high watermark (trim all)
```

The command prompts for confirmation when a deletion request is issued:
```
$ rpk topic trim-prefix test --offset 50 
TOPIC  PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test   0          48                  50

? Confirm trimming of the above offsets? Yes

TOPIC  PARTITION  NEW-START-OFFSET  ERROR
test   0          50
```

You may use a text/JSON/YAML file to trim multiple topics/offsets/partitions at the same time:
_This example contains an example of error handling when multiple requests are done at the same time_
```
$ cat /tmp/asJson.json
[
    {
        "topic":"test",
        "partition":0,
        "offset":50
    }, 
    {
        "topic":"test-2",
        "partition":0,
        "offset":7
    }
]

$ rpk topic trim-prefix --from-file /tmp/asJson.json
TOPIC   PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test    0          50                  50
test-2  0          6                   7

? Confirm trimming of the above offsets? Yes

TOPIC   PARTITION  NEW-START-OFFSET  ERROR
test-2  0          7                 
test    0          -                 OFFSET_OUT_OF_RANGE: The requested offset is not within the range of offsets maintained by the server.
```

It also supports passing specific timestamps to the`--offset` flag:
```
# Producing to a specific A timestamp
$ echo "k0,v0,1669856400000" | rpk topic produce test -f '%k,%v,%d\n'
Produced to partition 0 at offset 48 with timestamp 1669856400000.

# Producing to a specific B timestamp
$ echo "k0,v0,1669860400000" | rpk topic produce test -f '%k,%v,%d\n'
Produced to partition 0 at offset 49 with timestamp 1669860400000.

# Trimming exactly at A, gives the offset of A
$ rpk topic trim test -o "@1669856400000"
TOPIC  PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test   0          48                  48

# Trimming between A and B, gives the offset of B
$ rpk topic trim test -o "@1669856400002"
TOPIC  PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test   0          48                  49

# Trimming to a recent timestamp where there are no offsets, gives you the high water mark
rpk topic trim test -o "@-2h"          
TOPIC  PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test   0          48                  50
```

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

### Features

* rpk: adds `rpk topic trim-prefix`, rpk support for the DeleteRecords request. Now you can trim records to a specific offset in the given topic/partition.

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+		}
+		return filteredOffset.Offset, nil
+	}
+	// Not @-prefixed offset
+	return strconv.ParseInt(offset, 10, 64)
```
Move this to the top so the rest of the function can be un-indented:
```go
if !strings,HasPrefix(offset, "@") {
    return ...
}
...unindented
```

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+		if err != nil {
+			return 0, err
+		} else if end {
+			return 0, errors.New("invalid offset '@end'; use '--to-hwm' if you want to trim all")
```
I think it might be ok to allow @end -- I was thinking to block -1 only because users shouldn't need to know special numbers

r-vasquez replied:
Changed to allow `--offset end` and removed `--to-hwm`

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+			return 0, err
+		} else if end {
+			return 0, errors.New("invalid offset '@end'; use '--to-hwm' if you want to trim all")
+		} else if now.Before(startAt) {
+			return 0, fmt.Errorf("invalid offset %q; rpk topic trim does not support a time onward,  use '--to-hwm' if you want to trim all", offset)
```
Same thought -- should be ok to trim to the future to truncate to hwm?

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+		if all {
+			parsedOffset = -1 // -1 maps to the partition's high watermark.
+		} else {
+			parsedOffset, err = parseOffsetFlag(ctx, adm, topic, offset, p)
```
This function issues a metadata request for _every_ individual partition, and calculates time for every partition. This should instead take an input set of topics/partitions and only do the calculation & metadata request once.

r-vasquez replied:
Assuming by `This function` refers to `parseOffsetFlag` function:

This is true only for @-prefixed offsets (for numeric offsets, it just converts from string to int) and when all partitions are selected (or no --partition flag was specified, which is a design choice); We issue a `kadm.ListOffsetsAfterMilli` which I think it's what you refer to metadata + calculation.

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+// parseOffsetFile parse the given file whether it's a json, yaml or
+// space-delimited text. File content: topic | partition | offset.
+func parseOffsetFile(fs afero.Fs, file string) (kadm.Offsets, error) {
+	parsed, err := out.ParseFileArray[struct {
```
I was thinking that if a person wanted to provide a two column file, they could -- with the two columns being partition & offset. So, if this Parse errors, parse again with a struct that does not have the topic field. The two column format would require a topic arg.

Also, if a person _does_ supply a topic arg _and_ a three-col format, the topic for each offset in the file must match the topic arg.

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+
+			var o kadm.Offsets
+			if fromFile != "" {
+				o, err = parseOffsetFile(fs, fromFile)
```
(see below -- here we would pass the first arg if it exists, the topic, to this function)

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+				out.MaybeDie(err, "unable to parse file %q: %v", fromFile, err)
+			} else {
+				if len(args) == 0 {
+					out.Die("Error: required 'topic' not set; Usage: 'rpk topic trim-prefix [TOPIC]'")
```
We usually don't have `Error: ` prefixes, nor quote the args, nor give usage like this?

r-vasquez replied:
Yeah, I thought about that, but I came up with this while testing.

This is to "mimic" how Cobra errors when there's a missing flag/param, e.g:

```
$ rpk topic alter-config
Error: requires at least 1 arg(s), only received 0
Usage:
  rpk topic alter-config [TOPICS...] --set key=value --delete key2,key3 [flags]
  
$ rpk acl user update asdf
Error: required flag(s) "new-password" not set
Usage:
...
```

I can also do the following:

```
out.Die("Error: required arg 'topic' not set\n%v", cmd.UsageString())
```

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+	cmd.Flags().StringVarP(&fromFile, "from-file", "f", "", "File of topic/partition/offset for which to trim offsets for")
+	cmd.Flags().BoolVar(&noConfirm, "no-confirm", false, "Disable confirmation prompt")
+
+	cmd.MarkFlagsMutuallyExclusive("to-hwm", "offset")
```
--to-hwm also cannot be used with --from-file

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+// flags. It checks if the topic exists and defaults to all partitions in the
+// case that partitions were not provided.
+func parseOffsetArgs(ctx context.Context, adm *kadm.Client, all bool, topic, offset string, partitions []int) (kadm.Offsets, error) {
+	// First we check if the topic exists
+	listed, err := adm.ListTopics(ctx, topic)
+	if err != nil {
+		return nil, fmt.Errorf("unable to list topic %q metadata: %v", topic, err)
+	}
+	listed.EachError(func(d kadm.TopicDetail) {
+		err = fmt.Errorf("unable to list the topic %q: %v", topic, d.Err)
+	})
+	if err != nil {
+		return nil, err
+	}
```
We can skip this and just rely on ListOffsetsAfterMilli -- if the topic is not in the response, the topic does not exist.

Basically, we need to minimize trips to the broker because that's the slow part

r-vasquez replied:
I also use this to fill the partition slice if there is none L143 (i.e: all partitions)

twmb replied:
Can this be inside the conditional on L140, then?

edit: actually see my comment below

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+func printDeleteRecordRequest(ctx context.Context, adm *kadm.Client, o kadm.Offsets) (rerr error) {
+	tw := out.NewTable("TOPIC", "PARTITION", "PRIOR-START-OFFSET", "NEW-START-OFFSET")
+	o.Each(func(offset kadm.Offset) {
+		listed, err := adm.ListStartOffsets(ctx, offset.Topic)
```
This needs to be done only once, above the Each.

r-vasquez replied:
Not really, you can have multiple Topics in kadm.Offsets (e.g: --from-file with a file that describes multiple topics)

twmb commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,279 @@
+		// -1 maps to the partition's high watermark, and we want to print that
+		// instead of the '-1' that might be confusing for the end user.
+		if offset.At == -1 {
+			end, err := adm.ListEndOffsets(ctx, offset.Topic)
```
Same thought here -- this needs to be done only once.

r-vasquez replied:
https://github.com/redpanda-data/redpanda/pull/11825#discussion_r1252237254

twmb replied:
See my reply above, but more importantly: I don't think this branch makes sense, actually. -1 is only used for _input_ -- it is not returned from ListStartOffsets unless there is literally no offset that exists for this partition. This branch can be removed.

r-vasquez replied:
I'll move the listEndOffset above.

> I don't think this branch makes sense, actually. -1 is only used for input

Correct. -1 is used for input to the DeleteRecords request. (L150 we set the `-1`)

> it is not returned from ListStartOffsets unless there is literally no offset that exists for this partition.

We are not evaluating what comes from ListStartOffsets here, we are evaluating if `offset.At` which comes from the request.

> This branch can be removed.

I don't think it's right to remove this unless I'm not seeing something. In Line 150 if the `--offset` flag is `--offset end` we set the kadm.Offset to be "-1" so the request to DeleteRecords will set the new low watermark to be the high watermark.

In this branch/conditional we are trying to print the request, this conditional is only to avoid displaying the new start offset as `-1`:

```
$rpk topic trim-prefix test --offset end 
TOPIC  PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test   0          12                  -1
```

And instead, show:

```
$rpk topic trim-prefix test --offset end 
TOPIC  PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test   0          12                  50
```

r-vasquez commented:
The latest force pushes:

- The test is now only using rpk for trimming records.
- Fixed one merge conflict and rebased with dev.
- Addressed some comments.

r-vasquez commented on tests/rptest/tests/delete_records_test.py:
```diff
@@ -361,44 +356,41 @@ def test_delete_records_empty_or_missing_topic_or_partition(self):
         missing_topic = "foo_bar"
         out_of_range_prefix = "OFFSET_OUT_OF_RANGE"
         unknown_topic_or_partition = "UNKNOWN_TOPIC_OR_PARTITION"
-        missing_partition = "partition was not returned"
```
I had to remove this since rpk outputs for an unknown topic and a missing partition is the same: `UNKNOWN_TOPIC_OR_PARTITION`

Cc: @graphcareful

graphcareful replied:
all good

graphcareful commented on tests/rptest/clients/rpk.py:
```diff
@@ -1213,3 +1220,41 @@ def describe_log_dirs(self):
+        ]
+
+        if len(partitions) > 0:
+            cmd += ["--partitions", ",".join([str(x) for x in partitions])]
```
sweet, wouldn't it be easier to use `--json` ? I guess you wouldn't be testing the expected output , is that ok? FWIW its not that much code here anyway so either way is fine

graphcareful commented on tests/rptest/tests/delete_records_test.py:
```diff
@@ -19,12 +19,12 @@
 from rptest.services.cluster import cluster
 from rptest.services.kgo_verifier_services import KgoVerifierConsumerGroupConsumer, KgoVerifierProducer
 from rptest.tests.redpanda_test import RedpandaTest
-from rptest.clients.kcl import KCL
```
sweet

graphcareful commented on src/go/rpk/pkg/cli/topic/trim.go:
```diff
@@ -0,0 +1,294 @@
+
+This command allows you to trim records from topics, to trim the topics Redpanda
+sets the LogStartOffset for partitions to the requested offset. All segments 
+whose max partition is before the requested offset are deleted, and any records 
```
replace:
```
All segments 
whose max partition is before the requested offset are deleted
```
with:
```
All segments whose base offset is less then the requested offset are deleted,
```

twmb commented:
Both failures are the new https://github.com/redpanda-data/redpanda/issues/11944
Unrelated, per @graphcareful
